client/src/controllers/BlockController.ts

```diff
@@ -0,0 +1,83 @@
+import { Controller } from '@hotwired/stimulus';
+import { object } from 'prop-types';
+
+declare global {
+  interface Window {
+    initBlockWidget?: (id: string) => void; // Declare the initBlockWidget function as optional on the Window interface
+    telepath: any;
+  }
+}
+
+/**
+ * Adds the ability to unpack a Telepath object and render it on the controlled element.
+ *
+ * @example
+ * <div
+ * id="{id}"
+ * data-block
+ * data-controller="w-block"
+ * data-w-block-data-value="{block_json}"
+ * data-w-block-initial-value="{value_json}"
+ * data-w-block-error-value="{error_json}">
+ * </div>
+ */
+
+export class BlockController extends Controller<HTMLElement> {
+  /**
+   * initial - Initial value for the block
+   * data - Unpacked Telepath data
+   * error - Block error (optional)
+   */
+  static values = {
+    initial: { type: object, default: undefined },
+    data: { type: object, default: undefined },
+    error: { type: object, default: undefined },
+  };
+
+  /** Packed Telepath data. */
+  declare dataValue: object;
+  declare errorValue: object;
+  declare initialValue: object;
+
+  connect() {
+    const telepath = window.telepath;
+    if (!telepath) {
+      throw new Error("Telepath doesn't exit");
+    }
+    const output = telepath.unpack(this.dataValue);
+    const element = this.element;
+    if (!element.id) {
+      // Throw an error here saying that the element needs an id. I think... Only of the block render code needs an id.
+      throw new Error('Block element needs an id');
+    }
+    const id = element.id;
+    output.render(this.element, id, this.initialValue, this.errorValue);
+    this.dispatch('ready', { detail: { ...output }, cancelable: false });
+  }
+
+  static afterLoad() {
```

Comment:
@kituuu  I think you may misunderstand the purpose of the code in `afterLoad`, that code will run when the Stimulus controller will be registered.

This is not meant to use the NEW data attributes, but rather gives a way for the OLD (being removed) approach to be used by custom code. We are setting this up so that we can remove the equivalent code that existed in `client/src/entrypoints/admin/telepath/blocks.js` and it should 100% be exactly the same.

The reason we are 'moving' it to this new code is so that we can track it being removed fully once we bump to the next major release.

It also stops other code from trying to use this window global function which will be removed (either our own admin code or otherwise).

As such, we probably do not even need a unit test for the `afterLoad` code, it's not tested today and we will remove it in < a year so no need to add tests.

Reply:
Oh okay, so should I just revert the changes in `afterLoad` and remove the `BlockController.test.js`? @lb-

Reply:
yep. I have a suggested test coming with another review.